### PR TITLE
Reload waveform on track replacement in editor

### LINKED ISSUE #18713: Timeline waveform does not update on beatmap track changes
### Type

Cosmetic

### Bug description

As per the video, when the beatmap track is changed in editor, going into compose screen doesn't show the waveform in the timeline updated inline with the new track.

### Screenshots or videos

https://user-images.githubusercontent.com/22781491/173854533-46d636fb-6f1c-4624-a075-2bfff34efa7e.mp4

### Version

2022.612.0

### Logs

-

### PR DESCRIPTION
**Pull request by Feodor0090**
Close #18713.
The issue was in lazy storage of waveform in working beatmap - even when track is changed, cache was not reset, so always returned old waveform (or even null one). Couldn't think of anything better than to just expose a method for this.
7089bb6 is needed because the event fires on track reload, while previously used bindable doesn't.

### CODE REVIEW

File: osu.Game/Screens/Edit/Compose/Components/Timeline/Timeline.cs
Comment (bdach) [inline osu.Game/Screens/Edit/Compose/Components/Timeline/Timeline.cs:144]: Not sure I like `MusicController` being roped into this waveform reload flow... I'd much prefer that the change of the waveform itself could raise an event or something.

I'm not even really sure that this should be `WorkingBeatmap`'s responsibility to invalidate the waveform. There's an argument that `EditorBeatmap` could provide something like that just as well.

Comment (frenzibyte) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/Timeline.cs:144]: In fact, there's already a bindable for tracks in `EditorClock`, that should do the job.

Comment (Feodor0090) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/Timeline.cs]: > I'm not even really sure that this should be WorkingBeatmap's responsibility to invalidate the waveform.

Waveform creation requires file stream. Attempt to move creation code out of `WorkingBeatmap` will result in pure copy-pasting from `BeatmapManagerWorkingBeatmap`, i don't think it's good.

File: osu.Game/Beatmaps/WorkingBeatmap.cs
Comment (frenzibyte) [inline osu.Game/Beatmaps/WorkingBeatmap.cs:115]: I'm not 100% sure how safe it is to dispose a waveform from `WorkingBeatmap` itself, but nothing failed in game so it seems fine. There are no unmanaged/BASS resources to be freed by the disposal method, but I'm concerned about the GC handle causing the callbacks to be leaked on iOS, so I've pushed it.

**Review by frenzibyte — DISMISSED** [review]
Looks to work, but I'll wait for a second approval regarding the direction and especially the conversation below.

File: osu.Game/Screens/Edit/Timing/TapTimingControl.cs
Comment (bdach) [inline osu.Game/Screens/Edit/Timing/TapTimingControl.cs:192]: A bit dangerous this. If the track happens to change for whatever reason during load then this'll throw because of invalid thread for mutation. Probably best to move to `LoadComplete()` for safety.

**Review by bdach — DISMISSED** [review]
seems ok - a second approval from core team would go a long way but i'll probably merge in a day or two if that doesn't happen

File: osu.Game/Beatmaps/WorkingBeatmap.cs
Comment (peppy) [reply osu.Game/Beatmaps/WorkingBeatmap.cs:115]: Seems okay to go with until it fails.

File: osu.Game/Screens/Edit/Timing/TapTimingControl.cs
Comment (peppy) [inline osu.Game/Screens/Edit/Timing/TapTimingControl.cs:41]: Looking back at this, it's pretty weird that we're binding through `EditorClock`.... but I guess that's for another discussion another day.

File: osu.Game/Screens/Edit/Timing/TapTimingControl.cs
Comment (peppy) [inline osu.Game/Screens/Edit/Timing/TapTimingControl.cs:196]: I also find this super weird, that this is done here and not inside `WaveformComparisonDisplay` itself. Like, it's not getting the track passed in but if this line didn't exist, it'd likely fall over and fail due to an old waveform being disposed around it?

File: osu.Game/Screens/Edit/Timing/TapTimingControl.cs
Comment (bdach) [reply osu.Game/Screens/Edit/Timing/TapTimingControl.cs:41]: What would the alternative be? Via `WorkingBeatmap` directly? Make the waveform a bindable or something?

File: osu.Game/Screens/Edit/Timing/TapTimingControl.cs
Comment (bdach) [reply osu.Game/Screens/Edit/Timing/TapTimingControl.cs:196]: This is true and I kinda ignored that without looking into it more. On a fifteen-minute try it *should* be possible to make `WaveformComparisonDisplay` refresh itself, but it's not a five-minute job. May require some restructuring inside the class to work correctly as there are multiple pieces to update.

Comment (Feodor0090) [reply osu.Game/Screens/Edit/Timing/TapTimingControl.cs]: Everything is recalculated each beat pass / mouse movement, so looks like it's enough to just assign new waveforms to rows. In manual testing nothing seems broken.

**Review by peppy — APPROVED** [review]

**Pull request merged**
